app/models/user.rb

```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  before_save { email.downcase! }
+  validates :first_name,  presence: true, length: { maximum: 50 }
+  validates :last_name,  presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 }, format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, presence: true, length: { minimum: 6 }
```

Comment:
Please correct me if I'm wrong because I know nothing about Ruby. Is this the actual model? There seems to be a lot of parsing/format validation logic. From my experience I've usually seen this kind of formatting logic applied before an object is constructed and then inserted into the DB

Reply:
Nevermind. The schema is in another file. So is this user.rb essentially Ruby's equivalent of declaring a `public void insertNewUser(string firstName, string lastName, string Email...` kind of thing?

Reply:
This file is the Ruby model of a user. Think of it as a class definition, and we can instantiate and save User objects. The `validates` functions are used to validate that the objects created satisfy those criteria. Later, we will be adding functions to this class that these User objects can have and manipulate this data.